### PR TITLE
Add vdot to PyOpenCLArrayContext

### PR DESCRIPTION
**Pull request by alexfikl**
Add a vdot that works on array containers.

Inspired by wanting to remove some pyopencl specific code from [pytential](https://github.com/inducer/pytential/blob/2be5dd2cd53aa416badfaf87ceb0ae236bee2a1f/pytential/solve.py#L43-L54).

### CODE REVIEW

File: test/test_arraycontext.py [hunk 1]
```diff
@@ -219,9 +222,12 @@ def assert_close_to_numpy_in_containers(actx, op, args):
...
             make_obj_array([arg]) if isinstance(arg, DOFArray) else arg
             for arg in dofarray_args]
-    obj_array_result = actx.to_numpy(op(actx.np, *obj_array_args)[0][0])
 
-    assert np.allclose(obj_array_result, ref_result)
+    obj_array_result = op(actx.np, *obj_array_args)
+    if isinstance(obj_array_result, np.ndarray):
+        obj_array_result = obj_array_result[0][0]
+
+    assert np.allclose(actx.to_numpy(obj_array_result), ref_result)
```
Comment (inducer) [inline test/test_arraycontext.py:230]: Do any of the tests exercise complex-valued inputs?

**Review by inducer — COMMENTED** [review]
Thanks!

File: test/test_arraycontext.py (hunk 1 above)
Comment (alexfikl) [reply test/test_arraycontext.py:246]: Extended some of the tests to also check complex inputs in 5d67940.

File: arraycontext/impl/pyopencl/fake_numpy.py [hunk 2]
```diff
@@ -162,6 +162,27 @@ def _rec_ravel(a):
...
+
+        def _rec_vdot(xi, yi):
+            if is_array_container(xi):
+                assert type(xi) == type(yi)
+                return sum(_rec_vdot(subxi, subyi)
+                    for (_, subxi), (_, subyi) in zip(
+                        serialize_container(xi), serialize_container(yi)
+                    ))
+            else:
+                result = cl_array.vdot(xi, yi,
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply arraycontext/impl/pyopencl/fake_numpy.py:177]: We'll definitely need it when extending all the other reductions to array containers. Something like:
```
def rec_reduce_array_container(reduce_op, array_op, container):
	pass

def rec_multireduce_array_container(reduce_op, array_op, *args):
	pass
```
? I can work on that, but different PR maybe?

[the remark this answers is not in the document]
Comment (inducer) [reply arraycontext/impl/pyopencl/fake_numpy.py:177]: Sure, that's fine by me.

[the remark this answers is not in the document]
Comment (inducer) [reply arraycontext/impl/pyopencl/fake_numpy.py:177]: #61

**Comment by inducer** [comment]
Thanks!

**Pull request merged**